src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs

```diff
@@ -4,58 +4,203 @@
+            foreach (var set in actionSets)
+            {
+                foreach (var suggestedAction in set.Actions)
+                {
+                    // Filter out code actions with options since they'll show dialogs and we can't remote the UI and the options.
+                    if (suggestedAction.CodeAction is CodeActionWithOptions)
+                    {
+                        continue;
+                    }
+
+                    codeActions.Add(GenerateVSCodeAction(
+                        request, GetNestedActionsFromActionSet(suggestedAction),
+                        GetCodeActionKindFromSuggestedActionCategoryName(set.CategoryName!)));
+                }
+            }
+
+            return codeActions.ToArray();
+        }
+
+        private static VSCodeAction GenerateVSCodeAction(
+            CodeActionParams request,
+            CodeAction codeAction,
+            CodeActionKind codeActionKind,
+            string currentTitle = "")
+        {
+            using var _ = ArrayBuilder<VSCodeAction>.GetInstance(out var nestedActions);
+
+            if (!string.IsNullOrEmpty(currentTitle))
+            {
+                // Adding a delimiter for nested code actions, e.g. 'Suppress or Configure issues|Suppress IDEXXXX|in Source'
+                currentTitle += '|';
+            }
+
+            currentTitle += codeAction.Title;
+
+            // Nested code actions' unique identifiers consist of: parent code action unique identifier + '|' + title of code action
+            foreach (var action in codeAction.NestedCodeActions)
+            {
+                nestedActions.Add(GenerateVSCodeAction(request, action, codeActionKind, currentTitle));
+            }
+
+            // WPF uses the first underscore to create keyboard shortcuts, so we have to insert an extra underscore
+            // wherever there's an underscore for it to be displayed properly.
+            var title = codeAction.Title.Replace("_", "__");
+
+            return new VSCodeAction
+            {
+                Title = title,
+                Kind = codeActionKind,
+                Diagnostics = request.Context.Diagnostics,
+                Children = nestedActions.ToArray(),
+                Data = new CodeActionResolveData(currentTitle, request.Range, request.TextDocument)
+            };
+        }
+
+        /// <summary>
+        /// Get, order, and filter code actions.
+        /// </summary>
+        /// <remarks>
+        /// Used by CodeActionResolveHandler and RunCodeActionHandler.
+        /// </remarks>
+        public static async Task<ImmutableArray<CodeAction>> GetCodeActionsAsync(
+            Document document,
+            ICodeFixService codeFixService,
+            ICodeRefactoringService codeRefactoringService,
+            IThreadingContext threadingContext,
+            LSP.Range selection,
+            CancellationToken cancellationToken)
+        {
+            var actionSets = await GetActionSetsAsync(
```

Comment:
Looks like there's some duplicate logic between this and GetVSCodeActionsAsync, can we de-dupe it in some way?

Reply:
I think the only way to de-dupe would be by creating a generic helper method. But I believe that would raise another issue since the functions to create VSCodeActions and local Roslyn code actions are very different. We'd have to pass in those functions to the generic method, but each function takes in a different number/type of parameters. I think it'd be possible to de-dupe, but I believe it would be messy so I'm not sure if the benefits would outweigh the costs.